src/kernel.cpp

```diff
@@ -17,120 +17,10 @@
-
-        switch (arg.kind) {
-        case kernel_argument_kind::buffer:
-            dt = VK_DESCRIPTOR_TYPE_STORAGE_BUFFER;
-            break;
-        case kernel_argument_kind::buffer_ubo:
-            dt = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER;
-            break;
-        case kernel_argument_kind::ro_image:
-            dt = VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE;
-            break;
-        case kernel_argument_kind::wo_image:
-            dt = VK_DESCRIPTOR_TYPE_STORAGE_IMAGE;
-            break;
-        case kernel_argument_kind::sampler:
-            dt = VK_DESCRIPTOR_TYPE_SAMPLER;
-            break;
-        case kernel_argument_kind::local:
-            continue;
-        case kernel_argument_kind::pod:
-        case kernel_argument_kind::pod_ubo:
-            if (!pod_found) {
-                if (arg.kind == kernel_argument_kind::pod) {
-                    dt = VK_DESCRIPTOR_TYPE_STORAGE_BUFFER;
-                } else if (arg.kind == kernel_argument_kind::pod_ubo) {
-                    dt = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER;
-                }
-
-                m_pod_descriptor_type = dt;
-
-                pod_found = true;
-            } else {
-                continue;
-            }
-        }
-
-        VkDescriptorSetLayoutBinding binding = {
-            arg.binding,                 // binding
-            dt,                          // descriptorType
-            1,                           // decriptorCount
-            VK_SHADER_STAGE_COMPUTE_BIT, // stageFlags
-            nullptr                      // pImmutableSamplers
-        };
-
-        layoutBindings.push_back(binding);
-        smap[binding.descriptorType]++;
-    }
-
-    num_resources = layoutBindings.size();
-
-    if (!build_descriptor_set_layout(vkdev, layoutBindings)) {
-        return false;
-    }
-
-    return true;
-}
-
-bool cvk_kernel::build_descriptor_sets_layout_bindings_for_literal_samplers(
-    VkDevice vkdev, binding_stat_map& smap) {
-
-    std::vector<VkDescriptorSetLayoutBinding> layoutBindings;
-    for (auto& desc : m_program->literal_sampler_descs()) {
-        VkDescriptorSetLayoutBinding binding = {
-            desc.binding,                // binding
-            VK_DESCRIPTOR_TYPE_SAMPLER,  // descriptorType
-            1,                           // decriptorCount
-            VK_SHADER_STAGE_COMPUTE_BIT, // stageFlags
-            nullptr                      // pImmutableSamplers
-        };
-        layoutBindings.push_back(binding);
-        smap[binding.descriptorType]++;
-    }
-
-    if (!build_descriptor_set_layout(vkdev, layoutBindings)) {
-        return false;
-    }
-
-    return true;
-}
-
 std::unique_ptr<cvk_buffer> cvk_kernel::allocate_pod_buffer() {
     cl_int err;
-    auto buffer =
-        cvk_buffer::create(m_context, 0, m_pod_buffer_size, nullptr, &err);
+    auto buffer = cvk_buffer::create(
+        m_context, 0, m_entry_point->pod_buffer_size(), nullptr, &err);
```

Comment:
I'm wondering whether this shouldn't be delegated to `cvk_entry_point` since it owns the descriptors.

Reply:
I agree. I was going to explore moving this to `cvk_entry_point` as part of trying to fix issue #172, by having `cvk_entry_point` dish out suballocations of a single large buffer. I can do the simple step of delegating the current approach to `cvk_entry_point` as a first step in the PR though.